modules/viz/src/vizimpl.hpp

```diff
@@ -65,7 +65,8 @@ struct cv::viz::Viz3d::VizImpl
     void setOffScreenRendering();
 
     void removeAllLights();
-    void addLight(Vec3d position, Vec3d focalPoint, Color color, Color diffuseColor, Color ambientColor, Color specularColor);
+    void addLight(Vec3d position, Vec3d focalPoint, const Color &color, const Color &diffuseColor,
```

Comment:
Note that I cannot make `position` and `focalPoint` const reference in the implementation because after there are:
```
    light->SetPosition(position.val);
    light->SetFocalPoint(focalPoint.val);
```

which are defined as:
```
  // Description:
  // Set/Get the position of the light.
  // Note: The position of the light is defined in the coordinate
  // space indicated by its transformation matrix (if it exists).
  // Thus, to get the light's world space position, use
  // vtkGetTransformedPosition() instead of vtkGetPosition().
  vtkSetVector3Macro(Position,double);
  vtkGetVectorMacro(Position,double,3);
  void SetPosition(const float *a) {this->SetPosition(a[0],a[1],a[2]);};

  // Description:
  // Set/Get the point at which the light is shining.
  // Note: The focal point of the light is defined in the coordinate
  // space indicated by its transformation matrix (if it exists).
  // Thus, to get the light's world space focal point, use
  // vtkGetTransformedFocalPoint() instead of vtkGetFocalPoint().
  vtkSetVector3Macro(FocalPoint,double);
  vtkGetVectorMacro(FocalPoint,double,3);
  void SetFocalPoint(const float *a) {this->SetFocalPoint(a[0],a[1],a[2]);};
```

and

```
#define vtkSetVector3Macro(name,type) \
virtual void Set##name (type _arg1, type _arg2, type _arg3) \
  { \
  vtkDebugMacro(<< this->GetClassName() << " (" << this << "): setting " << #name " to (" << _arg1 << "," << _arg2 << "," << _arg3 << ")"); \
  if ((this->name[0] != _arg1)||(this->name[1] != _arg2)||(this->name[2] != _arg3)) \
    { \
    this->name[0] = _arg1; \
    this->name[1] = _arg2; \
    this->name[2] = _arg3; \
    this->Modified(); \
    } \
  }; \
virtual void Set##name (type _arg[3]) \
  { \
  this->Set##name (_arg[0], _arg[1], _arg[2]);\
  }

#define vtkGetVector3Macro(name,type) \
virtual type *Get##name () \
{ \
  vtkDebugMacro(<< this->GetClassName() << " (" << this << "): returning " << #name " pointer " << this->name); \
  return this->name; \
} \
virtual void Get##name (type &_arg1, type &_arg2, type &_arg3) \
  { \
    _arg1 = this->name[0]; \
    _arg2 = this->name[1]; \
    _arg3 = this->name[2]; \
  vtkDebugMacro(<< this->GetClassName() << " (" << this << "): returning " << #name " = (" << _arg1 << "," << _arg2 << "," << _arg3 << ")"); \
  }; \
virtual void Get##name (type _arg[3]) \
  { \
  this->Get##name (_arg[0], _arg[1], _arg[2]);\
  }
```

The error I get:
```
/home/opencv/modules/viz/src/vizimpl.cpp: In member function ‘void cv::viz::Viz3d::VizImpl::addLight(const Vec3d&, cv::Vec3d, const Color&, const Color&, const Color&, const Color&)’:
/home/opencv/modules/viz/src/vizimpl.cpp:215:33: error: invalid conversion from ‘const double*’ to ‘double*’ [-fpermissive]
     light->SetPosition(position.val);
                                 ^
In file included from /usr/include/vtk-6.2/vtkObject.h:43:0,
                 from /usr/include/vtk-6.2/vtkAlgorithm.h:33,
                 from /usr/include/vtk-6.2/vtkPolyDataAlgorithm.h:32,
                 from /usr/include/vtk-6.2/vtkAppendPolyData.h:32,
                 from /home/opencv-build/modules/viz/precomp.hpp:56:
/usr/include/vtk-6.2/vtkLight.h:101:3: note:   initializing argument 1 of ‘virtual void vtkLight::SetPosition(double*)’
   vtkSetVector3Macro(Position,double);
```

Reply:
@catree 
yes, you're right. I did not notice this.